File: module5-shared-memory/feedback_server.c
#include "mess_feedback.h"
#include <string.h>

int main() {
    // Create shared memory
    int shmid = shmget(SHM_KEY, sizeof(feedback_data_t), IPC_CREAT | 0666);
    if (shmid == -1) {
        perror("shmget");
        exit(1);
    }

    feedback_data_t *feedback = (feedback_data_t*)shmat(shmid, NULL, 0);
    if (feedback == (void*)-1) {
        perror("shmat");
        exit(1);
    }

    int semid = create_semaphore();

    // Initialize feedback data
    sem_wait(semid);
    feedback->good_count = 0;
    feedback->average_count = 0;
    feedback->poor_count = 0;
    feedback->total_feedback = 0;
    sem_signal(semid);

    printf("Mess Feedback Server initialized.\n");
    printf("Shared memory and semaphore created.\n");
    printf("Feedback counters reset to zero.\n");

    // Keep server running
    while (1) {
        sleep(5);
        sem_wait(semid);
        printf("Current Feedback - Good: %d, Average: %d, Poor: %d, Total: %d\n",
               feedback->good_count, feedback->average_count, 
               feedback->poor_count, feedback->total_feedback);
        sem_signal(semid);
    }

    shmdt(feedback);
    return 0;
}

File: module5-shared-memory/FeedbackServer.java
import java.util.concurrent.locks.ReentrantLock;
import java.util.concurrent.atomic.AtomicInteger;

public class FeedbackServer {
    // Shared memory simulation using static variables
    private static final AtomicInteger goodCount = new AtomicInteger(0);
    private static final AtomicInteger averageCount = new AtomicInteger(0);
    private static final AtomicInteger poorCount = new AtomicInteger(0);
    private static final AtomicInteger totalCount = new AtomicInteger(0);
    
    // Semaphore simulation using ReentrantLock
    private static final ReentrantLock lock = new ReentrantLock();
    
    public static void main(String[] args) {
        System.out.println("Mess Feedback Server started");
        System.out.println("Shared memory initialized - All counters reset to 0");
        
        // Keep server running and display status
        while (true) {
            try {
                Thread.sleep(5000);
                displayStatus();
            } catch (InterruptedException e) {
                break;
            }
        }
    }
    
    public static void submitFeedback(String type) {
        lock.lock();
        try {
            switch (type.toLowerCase()) {
                case "good":
                    goodCount.incrementAndGet();
                    break;
                case "average":
                    averageCount.incrementAndGet();
                    break;
                case "poor":
                    poorCount.incrementAndGet();
                    break;
                default:
                    System.out.println("Invalid feedback type: " + type);
                    return;
            }
            totalCount.incrementAndGet();
            System.out.println(type + " feedback submitted!");
        } finally {
            lock.unlock();
        }
    }
    
    public static String getStatus() {
        lock.lock();
        try {
            return String.format("{\"good\":%d,\"average\":%d,\"poor\":%d,\"total\":%d}",
                    goodCount.get(), averageCount.get(), poorCount.get(), totalCount.get());
        } finally {
            lock.unlock();
        }
    }
    
    private static void displayStatus() {
        lock.lock();
        try {
            System.out.printf("Current Feedback - Good: %d, Average: %d, Poor: %d, Total: %d%n",
                    goodCount.get(), averageCount.get(), poorCount.get(), totalCount.get());
        } finally {
            lock.unlock();
        }
    }
}